File: MPP/csharp/csharp/repository/AbstractDatabaseRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using csharp.domain;
using Npgsql;
using Serilog;
using Serilog.Core;

namespace csharp.repository;

public abstract class AbstractDatabaseRepository<ID, E> : IRepository<ID, E> where E : Entity<ID>
{
    protected readonly string url;

    public AbstractDatabaseRepository(string url)
    {
        this.url = url;
    }

    protected abstract E ExtractEntity(NpgsqlDataReader resultSet);

    protected abstract String GetFindOneQuery();

    protected abstract void SetIdParameter(NpgsqlCommand statement, ID id);
    
    public virtual E FindOne(ID id)
    {
        Log.Information("Se incearca conexiunea cu baza de date");
        using (NpgsqlConnection connection = new NpgsqlConnection(url))
        {
            connection.Open();
            using (NpgsqlCommand statement = connection.CreateCommand())
            {
                statement.CommandText = GetFindOneQuery();
                SetIdParameter(statement, id);
                using (NpgsqlDataReader resultSet = statement.ExecuteReader())
                {
                    if (resultSet.Read())
                    {
                        Log.Information("S-a gasit entitatea cu id-ul: " + id);
                        return ExtractEntity(resultSet);
                    }
                }
            }
        }
        return null;
    }
    
    protected abstract string GetFindAllQuery();

    public virtual List<E> FindAll()
    {
        List<E> entities = new List<E>();
        Log.Information("Se incearca conexiunea cu baza de date");
        using (NpgsqlConnection connection = new NpgsqlConnection(url))
        {
            connection.Open();
            using (NpgsqlCommand statement = connection.CreateCommand())
            {
                statement.CommandText = GetFindAllQuery();
                using (NpgsqlDataReader resultSet = statement.ExecuteReader())
                {
                    while (resultSet.Read())
                    {
                        E entity = ExtractEntity(resultSet);
                        Log.Information("S-a gasit entitatea cu id-ul: " + entity.Id);
                        entities.Add(entity);
                    }
                }
            }
        }
        return entities;
    }
    
    protected abstract string GetSaveQuery();
    
    protected abstract void SetSaveQueryParameters(NpgsqlCommand statement, E entity);

    public virtual E Save(E entity)
    {
        Log.Information("Se incearca conexiunea cu baza de date");
        using (NpgsqlConnection connection = new NpgsqlConnection(url))
        {
            connection.Open();
            using (NpgsqlCommand statement = connection.CreateCommand())
            {
                statement.CommandText = GetSaveQuery();
                SetSaveQueryParameters(statement, entity);
                int rowsInserted = statement.ExecuteNonQuery();
                if (rowsInserted == 0)
                {
                    Log.Information("S-a gasit entitatea cu id-ul: " + entity.Id);
                    return entity;
                }
            }
        }
        return null;
    }
    
    protected abstract string GetDeleteQuery();

    public virtual E Delete(ID id)
    {
        Log.Information("Se incearca conexiunea cu baza de date");
        E entity = FindOne(id);
        if (entity != null)
        {
            using (NpgsqlConnection connection = new NpgsqlConnection(url))
            {
                connection.Open();
                using (NpgsqlCommand statement = connection.CreateCommand())
                {
                    statement.CommandText = GetDeleteQuery();
                    SetIdParameter(statement, id);
                    Log.Information("S-a sters entitatea cu id-ul: " + id);
                    statement.ExecuteNonQuery();
                }
            }
        }
        return entity;
    }
    
    protected abstract string GetUpdateQuery();
    
    protected abstract void SetUpdateQueryParameters(NpgsqlCommand statement, E entity);
    
    public virtual E Update(E entity)
    {
        Log.Information("Se incearca conexiunea cu baza de date");
        using (NpgsqlConnection connection = new NpgsqlConnection(url))
        {
            connection.Open();
            using (NpgsqlCommand statement = connection.CreateCommand())
            {
                statement.CommandText = GetUpdateQuery();
                SetUpdateQueryParameters(statement, entity);
                int rowsUpdated = statement.ExecuteNonQuery();
                if (rowsUpdated == 0)
                {
                    Log.Information("S-a updatat entitatea cu id-ul: " + entity.Id);
                    return entity;
                }
            }
        }

        return null;
    }

}

File: MPP/javafx/src/main/java/org/example/repository/AbstractDatabaseRepository.java
package org.example.repository;

import org.example.domain.Entity;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.*;
import java.util.*;


public abstract class AbstractDatabaseRepository<ID, E extends Entity<ID>> implements Repository<ID, E> {

    protected String url;

    protected String username;

    protected String password;

    protected static final Logger logger = LoggerFactory.getLogger(AbstractDatabaseRepository.class);

    public AbstractDatabaseRepository(Properties properties) {
        url = properties.getProperty("jdbc.url");
        username = properties.getProperty("jdbc.user");
        password = properties.getProperty("jdbc.pass");
    }

    protected abstract E extractEntity(ResultSet resultSet) throws SQLException;

    protected abstract String getFindOneQuery();

    protected abstract void setIdParameter(PreparedStatement statement, ID id) throws SQLException;

    @Override
    public Optional<E> findOne(ID id) {
        logger.info("Se incearca conexiunea la baza de date");
        try (Connection connection = DriverManager.getConnection(url, username, password);
             PreparedStatement statement = connection.prepareStatement(getFindOneQuery())
        ) {
            setIdParameter(statement, id);
            ResultSet resultSet = statement.executeQuery();
            if (resultSet.next()) {
                E entity = extractEntity(resultSet);
                logger.info("S-a gasit entitatea: " + entity);
                return Optional.of(entity);
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        return Optional.empty();
    }

    protected abstract String getFindAllQuery();

    @Override
    public List<E> findAll() {
        List<E> entities = new ArrayList<>();

        logger.info("Se incearca conexiunea la baza de date");
        try (Connection connection = DriverManager.getConnection(url, username, password);
             PreparedStatement statement = connection.prepareStatement(getFindAllQuery());
             ResultSet resultSet = statement.executeQuery()
        ) {

            while (resultSet.next()) {
                E entity = extractEntity(resultSet);
                logger.info("S-a gasit entitatea: " + entity);
                entities.add(entity);
            }
            return entities;

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    protected abstract String getSaveQuery();

    protected abstract void setSaveParameters(PreparedStatement statement, E entity) throws SQLException;

    @Override
    public Optional<E> save(E entity) {
        logger.info("Se incearca conexiunea la baza de date");
        try (Connection connection = DriverManager.getConnection(url, username, password);
             PreparedStatement statement = connection.prepareStatement(getSaveQuery())
        ) {
            setSaveParameters(statement, entity);
            int rowsInserted = statement.executeUpdate();
            if (rowsInserted == 1) {
                logger.info("S-a inserat entitatea: " + entity);
                return Optional.of(entity);
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        return Optional.empty();
    }


    protected abstract String getDeleteQuery();

    @Override
    public Optional<E> delete(ID id) {
        logger.info("Se incearca conexiunea la baza de date");
        Optional<E> entityOptional = findOne(id);

        entityOptional.ifPresent(entity -> {
            try (Connection connection = DriverManager.getConnection(url, username, password);
                 PreparedStatement statement = connection.prepareStatement(getDeleteQuery())
            ) {
                setIdParameter(statement, id);
                int rowsDeleted = statement.executeUpdate();
                if (rowsDeleted == 0) {
                    logger.info("S-a sters entitatea:" + entity);
                }

            } catch (SQLException e) {
                throw new RuntimeException(e);
            }
        });

        return entityOptional;
    }

    protected abstract String getUpdateQuery();

    protected abstract void setUpdateParameters(PreparedStatement statement, E entity) throws SQLException;

    @Override
    public Optional<E> update(E entity) {

        logger.info("Se incearca conexiunea la baza de date");
        try (Connection connection = DriverManager.getConnection(url, username, password);
             PreparedStatement statement = connection.prepareStatement(getUpdateQuery())
        ) {
            setUpdateParameters(statement, entity);
            int rowsUpdated = statement.executeUpdate();
            if (rowsUpdated == 0) {
                logger.info("S-a updatat entitatea :" + entity);
                return Optional.of(entity);
            }

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        return Optional.empty();
    }


}